src/main/java/com/example/onlinebookstore/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,94 @@
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto addToCart(AddToCartRequestDto requestDto, Long userId) {
+        Long bookId = requestDto.getBookId();
+        Book book = bookRepository.findById(bookId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_BOOK + bookId));
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_USER + userId));
+
+        ShoppingCart shoppingCartFromDb = shoppingCartRepository.findByUserId(userId)
+                .orElseGet(() -> {
+                    ShoppingCart shoppingCart = new ShoppingCart();
+                    shoppingCart.setUser(user);
+                    shoppingCartRepository.save(shoppingCart);
+                    return shoppingCart;
+                });
+
+        CartItem cartItem = new CartItem();
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCartFromDb);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDb.getCartItems().add(cartItem);
+
+        return shoppingCartMapper.toDto(shoppingCartFromDb);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getCartByUserId(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_CART + userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto deleteByCartItemId(Long cartItemId, Long userId) {
+        cartItemRepository.deleteById(cartItemId);
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findByUserId(userId).orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_CART + userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto updateByCartItemId(
+            Long cartItemId,
+            PutCartItemRequestDto requestDto,
+            Long userId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_ITEM + cartItemId));
+        cartItem.setQuantity(requestDto.getQuantity());
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findByUserId(userId).orElseThrow(() -> new EntityNotFoundException(
+                        EXP_MSG_CAN_NOT_FIND_CART + userId));
```

Comment:
You use the same search logic in three places for `shoppingCart` by user ID. I suggest you divide that logic into your own method.